tiledb/api/c_api/enumeration/enumeration_api.cc

```diff
@@ -0,0 +1,204 @@
+/**
+ * @file tiledb/api/c_api/enumeration/enumeration_api.cc
+ *
+ * @section LICENSE
+ *
+ * The MIT License
+ *
+ * @copyright Copyright (c) 2023 TileDB, Inc.
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ * @section DESCRIPTION
+ *
+ * This file defines C API functions for the enumeration section.
+ */
+
+#include "../../c_api_support/c_api_support.h"
+#include "enumeration_api_experimental.h"
+#include "enumeration_api_internal.h"
+
+namespace tiledb::api {
+
+capi_return_t tiledb_enumeration_alloc(
+    tiledb_datatype_t type,
+    uint32_t cell_val_num,
+    int ordered,
+    const void* data,
+    uint64_t data_size,
+    const void* offsets,
+    uint64_t offsets_size,
+    tiledb_enumeration_t** enumeration) {
+  ensure_output_pointer_is_valid(enumeration);
+
+  // The nullptr-ness of data and offsets are handled by the underlying
+  // Enumeration constructor. As this is non-trivial (though not very
+  // complicated) logic, its not replicated here.
+
+  auto datatype = static_cast<sm::Datatype>(type);
+
+  bool is_ordered = false;
+  if (ordered != 0) {
+    is_ordered = true;
+  }
+
+  try {
+    *enumeration = tiledb_enumeration_handle_t::make_handle(
+        datatype,
+        cell_val_num,
+        is_ordered,
+        data,
+        data_size,
+        offsets,
+        offsets_size);
+  } catch (...) {
+    *enumeration = nullptr;
+    throw;
+  }
```

Comment:
```suggestion
  *enumeration = tiledb_enumeration_handle_t::make_handle(
      static_cast<sm::Datatype>(type),
      cell_val_num,
      ordered != 0,
      data,
      data_size,
      offsets,
      offsets_size);
```

How about we inline the two variables and remove setting the pointer to `nullptr` in case of errors? The latter is not something we do in the other `alloc` functions.

Reply:
For variable inlining, I prefer to not do that in general because its easy to miss logic and extra lines of code are cheap.

For setting the output pointer to nullptr that's something we do quite often in our C API. The idea here is that if someone forgets to check for an error code in C (which is often enough) the error they get dereferencing a nullptr will be a lot more obvious than dereferencing whatever random memory location the passed in pointer references.